go/vt/vtadmin/cache/cache.go

```diff
@@ -0,0 +1,255 @@
+	BackfillQueueSize int `json:"backfill_queue_size"`
+	// BackfillEnqueueWaitTime is how long to wait when attempting to enqueue a
+	// backfill request before giving up.
+	BackfillEnqueueWaitTime time.Duration `json:"backfill_enqueue_wait_time"`
+}
+
+// Cache is a generic cache supporting background fills. To add things to the
+// cache, call Add. To enqueue a background fill, call EnqueueBackfill with a
+// Keyer implementation, which will be passed to the fill func provided to New.
+//
+// For example, to create a schema cache that can backfill full payloads (including
+// size aggregation):
+//
+//		var c *cache.Cache[BackfillSchemaRequest, *vtadminpb.Schema]
+//		c := cache.New(func(ctx context.Context, req BackfillSchemaRequest) (*vtadminpb.Schema, error) {
+//			// Fetch schema based on fields in `req`.
+//			// If err is nil, the backfilled schema will be added to the cache.
+//			return cluster.fetchSchema(ctx, req)
+//		})
+//
+type Cache[Key Keyer, Value any] struct {
+	cache *cache.Cache
+
+	m        sync.Mutex
+	lastFill map[string]time.Time
+
+	ctx       context.Context
+	cancel    context.CancelFunc
+	wg        sync.WaitGroup
+	backfills chan *backfillRequest[Key]
+
+	fillFunc func(ctx context.Context, k Key) (Value, error)
+
+	cfg Config
+}
+
+// New creates a new cache with the given backfill func. When a request is
+// enqueued (via EnqueueBackfill), fillFunc will be called with that request.
+func New[Key Keyer, Value any](fillFunc func(ctx context.Context, req Key) (Value, error), cfg Config) *Cache[Key, Value] {
+	if cfg.BackfillEnqueueWaitTime <= 0 {
+		log.Warningf("BackfillEnqueueWaitTime (%v) must be positive, defaulting to %v", cfg.BackfillEnqueueWaitTime, DefaultBackfillEnqueueWaitTime)
+		cfg.BackfillEnqueueWaitTime = DefaultBackfillEnqueueWaitTime
+	}
+
+	if cfg.BackfillRequestTTL <= 0 {
+		log.Warningf("BackfillRequestTTL (%v) must be positive, defaulting to %v", cfg.BackfillRequestTTL, DefaultBackfillRequestTTL)
+		cfg.BackfillRequestTTL = DefaultBackfillRequestTTL
+	}
+
+	c := &Cache[Key, Value]{
+		cache:     cache.New(cfg.DefaultExpiration, cfg.CleanupInterval),
+		lastFill:  map[string]time.Time{},
+		backfills: make(chan *backfillRequest[Key], cfg.BackfillQueueSize),
+		fillFunc:  fillFunc,
+		cfg:       cfg,
+	}
+
+	c.ctx, c.cancel = context.WithCancel(context.Background())
+
+	c.wg.Add(1)
+	go c.backfill() // TODO: consider allowing N backfill threads to run, configurable
+
+	return c
+}
+
+// Add adds a (key, value) to the cache directly, following the semantics of
+// (github.com/patrickmn/go-cache).Cache.Add.
+func (c *Cache[Key, Value]) Add(key Key, val Value, d time.Duration) error {
+	return c.add(key.Key(), val, d)
+}
+
+func (c *Cache[Key, Value]) add(key string, val Value, d time.Duration) error {
+	c.m.Lock()
+	defer c.m.Unlock()
+
+	// Record the time we last cached this key, to check against
+	c.lastFill[key] = time.Now().UTC()
+	// Then cache the actual value.
+	return c.cache.Add(key, val, d)
```

Comment:
I would keep `c.cache.Add` outside of the locking for `c.m`. It adds unnecessary contention.